docs/reference/triggers/kafka.md

```diff
@@ -310,13 +312,20 @@ context.platform.on_signal(callback)
 * The call to the `explicit_ack()` method must be awaited, meaning the handler must be an async function, or provide an event loop to run that method. e.g.:
 ```py
 import asyncio
-import nuclio
+import nuclio_sdk
 
 def handler(context, event):
-  qualified_offset = nuclio.QualifiedOffset.from_event(event)
-  loop = asyncio.get_event_loop()
-  loop.run_until_complete(context.platform.explicit_ack(qualified_offset)
-  return "acked"
+    qualified_offset = nuclio_sdk.QualifiedOffset.from_event(event)
+  
+    loop = asyncio.get_event_loop()
+    if loop.is_running():
+        # add the coroutine to the loop
+        asyncio.run_coroutine_threadsafe(context.platform.explicit_ack(qualified_offset), loop)
```

Comment:
From my testing, this doesn't work. Does it work for you?

Reply:
I tested it with the changes in this current PR and it does work.

Reply:
I ran the following test code and explicit_ack was *not* called. Is there anything in your branch that would change that?
```python
import asyncio

async def explicit_ack(x):
    print(f"Called explicit_ack! with {x}")

def handler(context, event):
    qualified_offset = "placeholder"
  
    loop = asyncio.get_event_loop()
    if loop.is_running():
        print("loop is running")
        # add the coroutine to the loop
        asyncio.run_coroutine_threadsafe(explicit_ack(qualified_offset), loop)
    else:
        print("loop is not running")
        # run the loop and wait for the coroutine to finish
        loop.run_until_complete(explicit_ack(qualified_offset))
    
    return "acked"
```

Reply:
I did some research on this, and I think I understand why this might work at times and not work at other times.

Basically, `run_coroutine_threadsafe` submits the task to the event loop while the event loop is still busy running the handler (the very code that submits the new task).

So, this will not run `z()`:
```python
import asyncio


async def z():
    print("z")


def y():
    print("y")
    loop = asyncio.get_running_loop()
    future = asyncio.run_coroutine_threadsafe(z(), loop)


async def x():
    print("x")
    y()
    # await asyncio.sleep(0)


asyncio.run(x())
```
However, if you uncomment `await asyncio.sleep(0)`, which is basically a yield to the event loop, `z()` does indeed run. But I don't think this is something we can rely on (as demonstrated by the example handler function).

Reply:
I think the bottom line here is that the handler must either be async, or it has to call `explicit_ack` from another thread, which is what storey currently does. This raises another question though, which is whether this is thread safe. I think it's not, because of the risk of concurrent writes to the same socket (by explicit ack and by the wrapper itself).